### PR TITLE
rust: cross build (&template cleanup)

### PR DESCRIPTION
**Pull request by Cogitri**
Continuation of #3101. @Gottox currently seems to be very busy,
so I offered to get this working in #xbps earlier. Currently
building x86_64 -> aarch64, which doesn't work as of now, but early progresses somewhat further than before. However,
there's still stuff to be done:

- [x] Test if other arches actually build
- [ ] Test if the cross built binaries actually work
- [x] Adjust #2837 to work if cross (should be easy enough)

### CODE REVIEW

**Comment by maxice8** [comment]
needs to add [ci skip]

**Comment by Cogitri** [comment]
Currently tries to link against the host's LLVM: https://gist.github.com/0d9460d77cd0ce964b53f9526c6f47fe

**Comment by jnbr** [comment]
Here is a workaround for linking against host llvm:

```
--- rustc-1.28.0-src/src/librustc_llvm/build.rs.orig    2018-10-01 04:00:15.481334857 +0200
+++ rustc-1.28.0-src/src/librustc_llvm/build.rs 2018-10-01 04:01:15.145790417 +0200
@@ -227,8 +227,8 @@
             println!("cargo:rustc-link-search=native={}", &lib[9..]);
         } else if is_crossed {
             if lib.starts_with("-L") {
-                println!("cargo:rustc-link-search=native={}",
-                         lib[2..].replace(&host, &target));
+                println!("cargo:rustc-link-search=native={}{}",
+                         env::var("XBPS_CROSS_BASE").unwrap(), &lib[2..]);
             }
         } else if lib.starts_with("-l") {
             println!("cargo:rustc-link-lib={}", &lib[2..]);
```
The original `replace()` doesn't work because the vlaue of `host` doesn't appear in the output of `llvm-config --link-shared --ldflags`.

**Comment by Cogitri** [comment]
Thanks, that worked. Now it tries to run the just cross compiled stage0 compiler though (which obviously doesn't work due to it being compiled for a different arch). Will look into it.

**Comment by Gottox** [comment]
I had the approach to unset the cross build environment variables through stage0 and let the build system figure out which vars to set. An alternative approach that would safe build time is to skip stage0 and stage1 and directly build the final compiler. The bootstrapping is already done in the native build.

**Comment by Cogitri** [comment]
> An alternative approach that would safe build time is to skip stage0 and stage1 and directly build the final compiler.

Sounds good, as we can't run any stage1 binaries during cross compilation anyway

**Comment by Gottox** [comment]
But I have no idea how to convince the build system to start with stage2.

**Comment by Cogitri** [comment]
Seems like we've mixed up `build` and `host` though, if I understand it correctly `build` is the machine running the build (and the arch for which the stage0 compiler is compiled for) while `host` and `target` are the triplets we want to compile for.

**Comment by Cogitri** [comment]
Alright, cross building to aarch64 works now, testing the other arches. Does someone have aarch64 hardware to test the package? Otherwise I'll setup void on my RPI 3.

File: srcpkgs/rust/template [hunk 1]
```diff
@@ -65,81 +93,92 @@ post_extract() {
...
+			--set=target.${_build_triplet}.cc=${CC_host}
+			--set=target.${_build_triplet}.cxx=${CXX_host}
+			--set=target.${_build_triplet}.ar=${AR_host}
+			--set=target.${_build_triplet}.linker=${LD_host}
+		"
+
+		# Set 'llvm-config' for the cross target (host) so that we don't build
+		# LLVM for it again.
+		configure_args+="
+			--set=target.${_host_triplet}.llvm-config="/usr/bin/llvm-config"
```
Comment (jnbr) [inline srcpkgs/rust/template]: these nested quotes look odd

Comment (Cogitri) [reply srcpkgs/rust/template]: Hm, they do work though, although I guess they're not really required.

**Comment by Cogitri** [comment]
Do note that bootstrapping rust on the host platform (not cross compiling) doesn't work right now, but the issue that causes it to fail is fixed in 1.28->1.29 anyway, so I don't really feel like wasting time on fixing that, I just need the 1.28 rust musl dist tarball to be uploaded my someone.

**Comment by Cogitri** [comment]
Alright, rust builds for all platforms. Just adjusted the cargo build style and testing cross build for all of those packages.

File: srcpkgs/rust/template [hunk 2]
```diff
@@ -65,81 +93,92 @@ post_extract() {
...
+		--disable-full-bootstrap
 		--release-channel=stable
 		--disable-rpath
 		--disable-docs
 		--disable-codegen-tests
-		--enable-ccache
 		--llvm-root=/usr
-		--enable-local-rust
-		--local-rust-root=$wrksrc/stage0
+		--enable-local-rebuild
```
Comment (Cogitri) [inline srcpkgs/rust/template]: The problem with `--enable-local-rebuild` is that we'll need the a stage0 (or a version installed via xbps) of the same version as the rust version we want to bootstrap. This isn't problematic for everything but x86_64-musl, as we have to compile the stage0 tarball for that ourselves. That means that would have to cross compile rust from x86_64 -> x86_64-musl for the initial stage0 tarball (or compile that manually).

Comment (Gottox) [reply srcpkgs/rust/template]: Would it be a problem not to set local-rebuild then?

Comment (Cogitri) [reply srcpkgs/rust/template]: Ah, no, that works too, the comment that was previously in the rust template about local-rebuild led me to believe that we needed that.

**Pull request merged**